### PR TITLE
Encoder-decoder models: move embedding scale to nn.Module

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

This PR moves embedding scale to nn.Module in encoder decoder models, so that users who want to pass in `inputs_embeds` to the `forward` will get the same results as if they passed `input_ids` via `model.get_input_embeddings()(input_ids)`

The generation from embeds is not supported for these models, that's why we did not see the inconsistency. I think we do not need specific tests and here I can add support for generation from embeds in another PR if needed.

All the tests (+slow) for the changes models are passing on my end

### CODE REVIEW

**Review by gante — APPROVED** [review]
LGTM, thank you for fixing 🙌 

There are more models where this inconsistency happens (e.g. MVP, NllbMoe, ...), would you be able to propagate the pattern?

**Comment by gante** [comment]
@zucchini-nlp also, can you:
1. Check if there is a common test checking that the output of `forward` with `input_ids=input_ids` == output of forward with `inputs_embeds=model.get_input_embeddings()(input_ids)`
2. If there is not, can you add it? 🙏

**Comment by zucchini-nlp** [comment]
Hmm, okay, I thought the test for generation with inputs-embeds is enough. There is no such test, yes. If that's needed I can add it and that can trigger handling other models/cases if there are peculiarities :(

File: tests/models/align/test_modeling_align.py
Comment (gante) [inline tests/models/align/test_modeling_align.py:172]: Instead of a per-model exception, I'd suggest adding something like this in the test:

```py
signature = inspect.signature(model.forward)
if "inputs_embeds" not in signature:
    self.skipTest("model does not accept `inputs_embeds`")
```

This way we don't need many test overwrites 😉

File: tests/models/bart/test_modeling_bart.py
Comment (gante) [inline tests/models/bart/test_modeling_bart.py:507]: In this case, where the model raises a `NotImplementedError`, we can add a try/catch in the test, skipping the test if this exception is caught

**Review by gante — COMMENTED** [review]
Sorry for giving you tons of extra work here 😅 The result will be very cool in the long run, though 💪

File: tests/models/align/test_modeling_align.py
Comment (zucchini-nlp) [reply tests/models/align/test_modeling_align.py:172]: Yes, that is already in the test, I just did not remove yet the skips. Some models simply do not have the attribute get_inputs_embeddings() and yeah, I see your suggestion below. That can work :)

**Comment by zucchini-nlp** [comment]
Seems like it's ready for core maintainer's review. Reverted the previous commit with skips and tested that checking signature is working for most cases. In cases where it does not work, model-specific test skipping is kept.

File: tests/models/bridgetower/test_modeling_bridgetower.py
Comment (gante) [inline tests/models/bridgetower/test_modeling_bridgetower.py:511]: Shouldn't this (and most other skips) be caught in the skip conditions of the main test? What's missing?

Comment (zucchini-nlp) [reply tests/models/bridgetower/test_modeling_bridgetower.py:511]: Some models have all the things implemented, but do not use "ipnuts_embeds" any where, even though accept it in forward. I dod not remove unused args like this due to BC but can be cleaned up in another PR if needed

Comment (gante) [reply tests/models/bridgetower/test_modeling_bridgetower.py:511]: IMO -- instead of adding custom test code to handle a silent failure, let's make the properly handle the failure (if the user passes `inputs_embeds` to `forward`, raise a `NotImplementedError`, which can be caught in the test)

You've opened a Pandora's box of work in this PR, but it will be great in the long run! 🤗

Comment (zucchini-nlp) [reply tests/models/bridgetower/test_modeling_bridgetower.py:511]: Hmm, so in cases when the model actually accepts but not uses inputs_embeds, we should remove "get_input_embeddings()" method? Or raise error inside forward after the input_embeds are passed?

I would like to entirely remove inputs_embeds from the forward argument list, but this leads to my question: if we should have the same list of arguments in all forwards in all models, regardless of usage? And will it break anything?

Comment (amyeroberts) [reply tests/models/bridgetower/test_modeling_bridgetower.py:511]: > if we should have the same list of arguments in all forwards in all models

Not necessarily e.g. vision models don't accept input_ids. All models which are grouped together should have a common subset of inputs, which enable a full forward pass of the model. Most important is that the inputs have standardised names e.g. we don't have `token_ids` for one model, `input_ids` for another and that their behaviour is consistent e.g. `input_embeds` "means" the same thing across models. 

In the case of text models, I'd say yes, if they accept input_ids, then they should also accept input_embeds, and then throw an error within the forward pass after input_embeds are passes.

Comment (zucchini-nlp) [reply tests/models/bridgetower/test_modeling_bridgetower.py]: Btw, now the "test_inputs_embeds" can also be cleaned-up and all model-specific skips removed. I can do it later in another PR :)

**Comment by zucchini-nlp** [comment]
Failing tests are passing for me locally, and are not related to the changes

File: tests/test_modeling_common.py
Comment (amyeroberts) [inline tests/test_modeling_common.py:2755]:
```suggestion
```

File: tests/test_modeling_common.py
Comment (amyeroberts) [inline tests/test_modeling_common.py:2757]: We shouldn't have try/except patterns in tests: either we're deliberately triggering an error or we're not. Raising exceptions is for within code, allowing us to elegantly handle errors at runtime. Instead, models which don't use `input_embeds` should explicitly skip this test, and the try/except block be removed

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing this - definitely a behaviour we want.

V. nicely handled and tested - just a few small things to address before merge

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:2756]: Hmm I thought we are adding all the NotImplementedErrors to get rid of many copies "skipTests", as discussed with @gante above:

> IMO -- instead of adding custom test code to handle a silent failure, let's properly handle the failure (if the user passes inputs_embeds to forward, raise a NotImplementedError, which can be caught in the test)

I can bring back all skips if that's needed for tests-consistency but if we are specifically catching only "NotImplementedErrors" isn't it okay?

File: tests/test_modeling_common.py
Comment (amyeroberts) [reply tests/test_modeling_common.py]: We shouldn't have any print statements in tests

File: tests/test_modeling_common.py
Comment (amyeroberts) [reply tests/test_modeling_common.py:2756]: Ah, OK, sorry, I missed the bit about handling in the tests. 

We should raise NotImplementedError on the model side, but let users handle that however they want, and then explicitly skip in the tests using `unittest.skip` for the specific models. This avoids accidentally skipping because a different NotImplementedError is raised. It's true we end up with more code, [but it's better for tests to be DAMP](https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html) i.e. very clear and explicit.

Comment (gante) [reply tests/test_modeling_common.py:2756]: > [but it's better for tests to be DAMP](https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html)

hehe the opposite of what I've been telling @zucchini-nlp in this PR 🙈  By bad :D

Comment (zucchini-nlp) [reply tests/test_modeling_common.py:2756]: Ah, okay, TIL about the DAMP principle

**Comment by zucchini-nlp** [comment]
@amyeroberts done, all the NotImplementedError models are skipped in their modeling_test files and the CI is green

**Review by amyeroberts — APPROVED** [review]
Beautiful - thanks for fixing this tricky issue and making sure it's well tested!

**Pull request merged**
